proposals/0010-stream-type.md

```diff
@@ -0,0 +1,162 @@
+      liveIndicatorEl.classList.remove('hide');
+      seekBarEl.classList.add('hide');
+    }
+    // For "on-demand", do the opposite.
+    // NOTE: In this example, if `mediaEl` has not implemented `streamType`,
+    // it will show the same UI as "on-demand".
+    else {
+      seekBarEl.classList.remove('hide');
+      liveIndicatorEl.classList.add('hide');
+    }
+    // Show the standard UI and hide the "big play button"
+    bigPlayButtonEl.classList.add('hide');
+    playerUIEl.classList.remove('hide');
+  }
+};
+
+// Update the UI whenever "streamtypechange" is dispatched.
+mediaEl.addEventListener('streamtypechange', updateUI);
+// Also update the UI whenever "emptied" is dispatched to account for unloading the media.
+mediaEl.addEventListener('emptied', updateUI);
+```
+
+# Reference-level explanation
+
+## Property: `streamType`
+
+A string value that represents the type of media stream that *__MUST__* be one of the values listed below.
+
+__NOTE__: Any computation of the `streamType` value that requires fetching of playlists, manifests, or the like, *__SHOULD__* respect the load and `preload` state specified on the extended `HTMLMediaElement` instance. For example, if `preload="none"`, implementors should wait until e.g. a `"loadstart"` event is dispatched from the instance.
+
+### Possible values
+
+- `undefined` - Unimplemented, in which case UI implementors may rely on a "best guess" solution for modeling the stream type (See *§Recommended inferred value from native `HTMLMediaElement`*, below).
+- `"on-demand"` - The current loaded media source is an On Demand Stream.
+- `"live"` - The current loaded media source is a Live Stream (including potentially "DVR" or "sliding window" live streams).
+- `"unknown"` - No media source is set, the source is not yet loaded (e.g. because `preload="none"`), or the stream type has yet to be determined.
+
+### Recommended computation for RFC8216 (aka HLS)
+
+`streamType = (#EXT-X-PLAYLIST-TYPE === "VOD") ? "on-demand" : "live"`
+
+__Context__:
+
+> A Media Playlist has further constraints on its updates if it contains an EXT-X-PLAYLIST-TYPE tag.  An EXT-X-PLAYLIST-TYPE tag with a value of VOD indicates that the Playlist file MUST NOT change.  An EXT-X-PLAYLIST-TYPE tag with a value of EVENT indicates that the Server MUST NOT change or remove any part of the Playlist file, with the exception of EXT-X-PART tags and Media Metadata tags as described above; the Server MAY append lines to the Playlist.
+
+  \- From: https://datatracker.ietf.org/doc/html/draft-pantos-hls-rfc8216bis-12#section-6.2.1
+
+#### NOTE: Why not `#EXT-X-ENDLIST` detection?
+
+The `#EXT-X-ENDLIST` tag only indicates that a particular playlist will no longer be updated with additional Media Segments or media segment parts (`#EXTINF` / `#EXT-X-PART-INF`). Using this value therefore conflates, e.g., an ended live stream with an on demand stream, and thus should not be relied upon for stream type detection.
```

Comment:
Based on the definition...

> On-demand: "Media content that is entirely available"
> Live: "and the end of the media...will only be available sometime in the future [i.e. not ended yet]

Should endlist detection trigger a change, at least for an Event/DVR playlist? This was at least the one clear justification I could come up with for having a specific event fire.

Reply:
> Should endlist detection trigger a change, at least for an Event/DVR playlist?

Hmmm I strongly think no, and if anything should change, it's word-smithing around the definitions.

Reply:
Doesn't the video element get a duration when endlist comes in? What might happen in a live UI that now has a duration? What's the expectation of what should happen in the live UI when a video is no longer live, DVR or otherwise?

>  it's word-smithing around the definitions

Not sure I follow that. Isn't this a decision of if the event should fire when a DVR live stream ends?

Reply:
Perhaps I am unclear on your proposal. Are you proposing that 

1. `"streamtypechange"` should be dispatched, 
2. the value of `streamType` itself should change (from `"live"` to `"event"`) based on a duration change/`EXT-X-ENDLIST`, or 
3. both?

I'm cool with ensuring that `"streamtypechange"` should fire whenever the stream type in fact has changed (aka (1) and what I've said in the other discussion thread). However, I don't believe the stream type should change as a result of a live stream ending (whether or not it's DVR) (aka (2)), as that conflates two concepts that many UI developers plausibly don't want to conflate ("a live stream that has now ended" vs. "this was a live stream and now it's an on demand stream").

Reply:
I think that `Why not simply rely on ``duration``?` section answers my concerns for the most part. `duration` could be used to know whether the stream is live *now* and `streamType` could be used to know whether this is a live stream (with the caveat that the PLAYLIST-TYPE/MPD@type doesn't change for this source).